src/Features/Core/Portable/IntroduceVariable/AbstractIntroduceParameterService.cs

```diff
@@ -0,0 +1,688 @@
+            }
+        }
+
+        private static int GetInsertionIndex(SemanticDocument document, Compilation compilation,
+            ISyntaxFactsService syntaxFacts, SyntaxNode methodDeclaration, CancellationToken cancellationToken)
+        {
+            var parameterList = syntaxFacts.GetParameterList(methodDeclaration);
+            Contract.ThrowIfNull(parameterList);
+            var symbol = document.SemanticModel.GetDeclaredSymbol(parameterList.GetRequiredParent(), cancellationToken);
+            Contract.ThrowIfNull(symbol);
+            var parameterSymbolList = ((IMethodSymbol)symbol).Parameters;
+            var insertionIndex = 0;
+
+            foreach (var parameterSymbol in parameterSymbolList)
+            {
+                if (!parameterSymbol.HasExplicitDefaultValue && !parameterSymbol.IsParams && !parameterSymbol.Type.Equals(compilation.GetTypeByMetadataName(typeof(CancellationToken)?.FullName!)))
+                {
+                    insertionIndex++;
+                }
+            }
+
+            return insertionIndex;
+        }
+
+        /// <summary>
+        /// Adds an argument which is an invocation of the newly created method to the callsites 
+        /// of the method invocations where a parameter was added
+        /// </summary>
+        private static SyntaxNode GenerateNewInvocationExpressionForTrampoline(ISyntaxFactsService syntaxFacts,
+            SyntaxEditor editor, SyntaxNode currentInvocation, SyntaxNode invocationExpression,
+            string newMethodIdentifier, int insertionIndex)
+        {
+            var invocationArguments = syntaxFacts.GetArgumentsOfInvocationExpression(currentInvocation);
+            var methodName = editor.Generator.IdentifierName(newMethodIdentifier);
+            var expressionFromInvocation = syntaxFacts.GetExpressionOfInvocationExpression(invocationExpression);
+            var newMethodInvocation = editor.Generator.InvocationExpression(methodName, invocationArguments);
+            var allArguments = invocationArguments.Insert(insertionIndex, newMethodInvocation);
+            return editor.Generator.InvocationExpression(expressionFromInvocation, allArguments);
+        }
+
+        /// <summary>
+        /// Goes through all of the invocations and replaces the expression with identifiers from the invocation 
+        /// arguments and rewrites the call site with the updated expression as a new argument
+        /// </summary>
+        private async Task<Solution> RewriteSolutionWithNewMethodAsync(
+            SemanticDocument semanticDocument, TExpressionSyntax expression, IMethodSymbol methodSymbol,
+            bool allOccurrences, string parameterName,
+            ImmutableDictionary<Document, List<TInvocationExpressionSyntax>> callSites, bool trampoline,
+            CancellationToken cancellationToken)
+        {
+            var firstCallSite = callSites.Keys.First();
+            var modifiedSolution = firstCallSite.Project.Solution;
```

Comment:
you don't need thsese lines.  you're already passsing in teh document.  So just grab the solution from that.

Reply:
removedf